### PR TITLE
Basic Parent Boss Controller Class

### PR DESCRIPTION
**Pull request by Vitality-Labs**

### CODE REVIEW

File: Assets/Scripts/BossController.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+public abstract class BossController : MonoBehaviour {
+    // Pubic Members
+    public float health = 500f; // Starting health value
+    public float movementSpeed = 8f; // Speed enemy will move through the world
+    public float attackRange = 20f; // Max distance you can be from player to attack
+    public float attackStrength = 15f; // Damage that will be done to player upon successful attack
+    public Transform playerObject; // Player transform object
+    
+    // Private Members
```
Comment (AnthonyReinman) [inline Assets/Scripts/BossController.cs:13]: private float _health; // Current health value

File: Assets/Scripts/BossController.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    private Rigidbody2D _body; // Enemy rigidbody, will get from object
+    private Vector2 _playerDirection; // Direction player is relative to player, used for tracking and following player
+    private float _playerDistance;
+
```
Comment (AnthonyReinman) [inline Assets/Scripts/BossController.cs:17]: // Property for Health
    public float Health {
        get { return _health; }
        set {
            _health = Mathf.Clamp(value, 0, maxHealth);
            OnHealthChanged();
            if (_health <= 0) {
                Die();
            }
        }
    }

File: Assets/Scripts/BossController.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    // Start is called once before the first execution of Update after the MonoBehaviour is created
+    void Start() {
+        _body = GetComponent<Rigidbody2D>();
+        if (!playerObject) {
+            playerObject = GameObject.FindGameObjectWithTag("Player").transform;
+            if (!playerObject) Debug.LogError("Enemy failed to find player!");
+        }
```
Comment (AnthonyReinman) [inline Assets/Scripts/BossController.cs:24]: // Initialize health
        Health = maxHealth;

File: Assets/Scripts/BossController.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public abstract class BossController : MonoBehaviour {
+    // Pubic Members
+    public float health = 500f; // Starting health value
```
Comment (AnthonyReinman) [inline Assets/Scripts/BossController.cs:7]: change to maxHealth

**Review by AnthonyReinman — APPROVED** [review]
Added maxHealth:
Replaced the public health field with maxHealth to define the maximum health value.

Added _health:
Private field _health to store the current health value.

Added Health Property:
Encapsulates the _health field with additional logic to clamp the value and trigger events.

Initialized Health in Start:
Set Health to maxHealth in the Start method to initialize the health value.
Added OnHealthChanged and Die Methods:

Placeholder methods for additional logic when health changes and when the boss dies.

With these changes we can adjust how the boss reacts when at certain thresholds. Let me know what you think. Should be easier to add additional logic if we want to based on health.

I'm good with leaving it how it is as well. Just something I saw recently when looking at controllers for enemies.

**Review by DreViez — APPROVED** [review]

**Pull request merged**